resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/kubernetes/NodeCacheManager.scala

```diff
@@ -0,0 +1,97 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.scheduler.cluster.kubernetes
+
+import java.util.concurrent.ExecutorService
+
+import scala.collection.JavaConverters._
+
+import io.fabric8.kubernetes.api.model.Pod
+import io.fabric8.kubernetes.client.{KubernetesClient, KubernetesClientException, Watch, Watcher}
+import io.fabric8.kubernetes.client.Watcher.Action
+import io.fabric8.kubernetes.client.internal.readiness.Readiness
+
+import org.apache.spark.SparkException
+import org.apache.spark.internal.Logging
+import org.apache.spark.util.ThreadUtils
+
+private[spark] class NodeCacheManager (
+    val client: KubernetesClient,
+    val dsNamespace: String,
+    val dsLabels: Map[String, String]) extends Logging {
+
+  private var shufflePodCache = scala.collection.mutable.Map[String, String]()
+  private var watcher: Watch = _
+
+  def start(): Unit = {
+    // seed the initial cache.
+    val pods = client.pods().withLabels(dsLabels.asJava).list()
+    for (pod <- pods.getItems.asScala) {
+      if (Readiness.isReady(pod)) {
+        addShufflePodToCache(pod)
+      } else {
+        logWarning(s"Found unready shuffle pod ${pod.getMetadata.getName} " +
+          s"on node ${pod.getSpec.getNodeName}")
+      }
+    }
+
+    watcher = client
+      .pods()
+      .withLabels(dsLabels.asJava)
+      .watch(new Watcher[Pod] {
+        override def eventReceived(action: Watcher.Action, p: Pod): Unit = {
+          action match {
+            case Action.DELETED | Action.ERROR =>
+              shufflePodCache.remove(p.getSpec.getNodeName)
+            case Action.ADDED | Action.MODIFIED if Readiness.isReady(p) =>
+              addShufflePodToCache(p)
+          }
+        }
+        override def onClose(e: KubernetesClientException): Unit = {}
+      })
+  }
+
+  private def addShufflePodToCache(pod: Pod): Unit = {
+    shufflePodCache.get(pod.getSpec.getNodeName) match {
+      case Some(registeredPodName) =>
+        logError(s"Ambiguous specification of shuffle service pod. " +
+          s"Found multiple matching pods: ${pod.getMetadata.getName}, " +
+          s"${registeredPodName} on ${pod.getSpec.getNodeName}")
+
+        throw new SparkException(s"Ambiguous specification of shuffle service pod. " +
+          s"Found multiple matching pods: ${pod.getMetadata.getName}, " +
+          s"${registeredPodName} on ${pod.getSpec.getNodeName}")
+
+      case nil =>
+        shufflePodCache(pod.getSpec.getNodeName) = pod.getStatus.getPodIP
+    }
+  }
+
+  def stop(): Unit = {
+    watcher.close()
+  }
+
+  def getShufflePodForExecutor(executorNode: String): String = {
+    shufflePodCache.get(executorNode) match {
+      case Some(pod) => pod
+      case _ =>
+        throw new SparkException(s"Unable to find shuffle pod on node $executorNode")
```

Comment:
Throwing the exception here should be fine, right? This is being processed on a separate thread in the RPC environment. Thus the exception here should only propagate to the executor that asked for this configuration.